usr/lib/ica_s390_stdll/ica_specific.c

```diff
@@ -4261,8 +4261,13 @@ CK_RV token_specific_ec_generate_keypair(STDLL_TokData_t *tokdata,
     /* Generate key data for this key object */
     rc = p_ica_ec_key_generate(ica_data->adapter_handle, eckey);
     if (rc != 0) {
-        TRACE_ERROR("ica_ec_key_generate() failed with rc=%d.\n", rc);
-        ret = CKR_FUNCTION_FAILED;
+        if (rc == EINVAL) {
```

Comment:
I guess libica may return EINVAL also for other reasons? With this we treat every EINVAL return as CKR_CURVE_NOT_SUPPORTED. However, without changing libica to return a more precise error indication in case of an unsupported curve, I don't see a better option to handle this.

Reply:
Yes, right. We would have to introduce a specific error code for 'curve not supported by openssl'. In case there is no card, libica returns ENODEV, but only if fallbacks are disabled. With fallbacks enabled, we get either 0 (=ok), EINVAL (=curve not supported by openssl), EACCES (fips error), or EIO (any other error). 
We could introduce EPERM, like 'operation not permitted by openssl' for unsupported curves, but would break the libica API with that. Currently, the libica book is not very specific here, but existing customer applications might check this rc.

Reply:
I guess this is a problem we will tackle once moving all sw-fallbacks from libica to the ica-token?

Reply:
Yes, exactly.

Reply:
@ifranzki :
> I guess libica may return EINVAL also for other reasons? 

@jschmidb 
> Yes, right. [...] EINVAL (=curve not supported by openssl) [...]

Sounds to me like EINVAL = curve not supported by openssl.. in what other cases EINVAL is returned ?

Anyway, if you need to change the return codes, please update the book and the comments in the code.

Reply:
> in what other cases EINVAL is returned ?

The usual parm checking, e.g. if any parm is null